src/zimscraperlib/zim/indexing_item.py

```diff
@@ -0,0 +1,199 @@
+
+    When indexdata (with title and content) is passed, this is used for indexing.
+
+    Otherwise, indexdata is automatically computed based on item content when possible.
+
+    For now only PDF documents are supported for automatic computation of index data,
+    and search index is populated only with PDF text content (i.e. no OCR is done on PDF
+    images)
+    """
+
+    def __init__(
+        self,
+        content: str | bytes | None = None,
+        fileobj: io.BytesIO | None = None,
+        filepath: pathlib.Path | None = None,
+        path: str | None = None,
+        title: str | None = None,
+        mimetype: str | None = None,
+        hints: dict | None = None,
+        indexdata: IndexData | None = None,
+        **kwargs: Any,
+    ):
+
+        super().__init__(
+            content, fileobj, filepath, path, title, mimetype, hints, **kwargs
+        )
+
+        if indexdata:
+            # indexdata has been externaly configured
+            self.get_indexdata = lambda: indexdata
+        else:
+            self.get_auto_title_and_index_data()
+
+    def get_auto_title_and_index_data(self):
+        """Populate item index data and title automatically from content"""
+
+        # content was set manually
+        content = getattr(self, "content", None)
+        if content is not None:
+            if not isinstance(content, (str, bytes)):
+                raise AttributeError(
+                    f"Unexpected type for content: {type(content)}"
+                )  # pragma: no cover
+            mimetype = get_content_mimetype(
+                content.encode() if isinstance(content, str) else content
+            )
+            if mimetype == "application/pdf":
+                index_data = get_pdf_index_data(content=content)
+            else:
+                return
+
+        # using a file-like object
+        fileobj = getattr(self, "fileobj", None)
+        if fileobj:
+            if not isinstance(fileobj, io.BytesIO):
+                raise AttributeError(
+                    f"Unexpected type for content: {type(fileobj)}"
+                )  # pragma: no cover
+            mimetype = get_content_mimetype(fileobj.getvalue())
+            if mimetype == "application/pdf":
+                index_data = get_pdf_index_data(fileobj=fileobj)
+            else:
+                return
+
+        # using a file path
+        filepath = getattr(self, "filepath", None)
+        if filepath:
+            if not isinstance(filepath, pathlib.Path):
+                raise AttributeError(
+                    f"Unexpected type for content: {type(filepath)}"
+                )  # pragma: no cover
+            mimetype = get_file_mimetype(filepath)
+            if mimetype == "application/pdf":
+                index_data = get_pdf_index_data(filepath=filepath)
+            else:
+                return
+
+        self.get_indexdata = lambda: index_data
+        self.title = index_data.title
```

Comment:
I think this implicit replacement might not be wanted. If we want to keep it auto, we should maybe only apply it if there's no title…

Reply:
Agreed